examples/decoding/plot_ssd_spatial_filters.py

```diff
@@ -28,144 +23,372 @@
+            # rephase for filtering
+            X_aux = np.reshape(inst, [n_epochs, n_channels * n_samples])
+            inst_noise = filter_data(
+                X_aux, self.sampling_freq, **self.filt_params_noise)
+            # subtract signal:
+            inst_noise -= inst_signal
+            # Estimate single trial covariance
+            # reshape to original shape
+            inst_signal = np.reshape(
+                inst_signal, [n_epochs, n_channels, n_samples])
+            # reshape to original shape
+            inst_noise = np.reshape(
+                inst_noise, [n_epochs, n_channels, n_samples])
+        return inst_signal, inst_noise
+
+    def fit(self, inst):
+        """Estimate the SSD decomposition on raw or epoched data.
+
+        Parameters
+        ----------
+        inst : instance of Raw data or array with shape
+            (n_epochs, n_channels, n_times)
+            The data on which to estimate the SSD.
+
+        Returns
+        -------
+        self : instance of SSD
+            Returns the modified instance.
+        """
+        if isinstance(inst, BaseRaw):
+            if self.picks_ is None:
+                raise ValueError('picks should be provided')
+            self.max_components = len(self.picks_)
+            inst_signal, inst_noise = self.filter_data(inst)
+            cov_signal = compute_raw_covariance(inst_signal, picks=self.picks_,
+                method=self.estimator, rank=self.rank)
+            cov_noise = compute_raw_covariance(inst_noise, picks=self.picks_,
+                method=self.estimator, rank=self.rank)
             del inst_noise
             del inst_signal
         else:
-            raise NotImplementedError()
-
-        self.eigvals_, self.filters_ = eigh(cov_signal.data, cov_noise.data)
+            if not isinstance(inst, np.ndarray):
+                raise NotImplementedError()
+            self._check_X(inst)
+            # data X is epoched
+            # part of the following code is copied from mne csp
+            n_epochs, n_channels, n_samples = inst.shape
+            self.max_components = n_channels
+            X_s, X_n = self.filter_data(inst)
+            covs = np.empty((n_epochs, n_channels, n_channels))
+            for ii, epoch in enumerate(X_s):
+                covs[ii] = _regularized_covariance(epoch, reg=self.estimator,
+                    method_params=self.cov_method_params, rank=self.rank)
+            cov_signal = covs.mean(0)
+            # Covariance matrix for the flanking frequencies (noise)
+            # Estimate single trial covariance
+            covs_n = np.empty((n_epochs, n_channels, n_channels))
+            for ii, epoch in enumerate(X_n):
+                covs_n[ii] = _regularized_covariance(epoch, reg=self.estimator,
+                    method_params=self.cov_method_params,
+                    rank=self.rank)
+            cov_noise = covs_n.mean(0)
+        eigvals_, eigvects_ = eigh(cov_signal.data, cov_noise.data)
```

Comment:
yes, I followed the csp implementation in that sense. But it can be fixed by adding the option and then just simply calling inst= epochs.get_data()[:, self.picks]

Reply:
Or you you define in each conditional a ``cov_sginal/noise`` variable and  remove the ``cov_sginal.data``.

Reply:
so, I added the option to work with epoch instance in fit, so it should work for raw, epoch or np.arrays.

Reply:
Not sure we're on the same page here. ```_regularized_covariance``` returns an array. Using the array's .data attribute is a bit unusual and many people may not even know it exists. The code, in the context of MNE, suggests you process Covariance objects. I'd just write a code that extracts the array from the Covariance object and looks like processing arrays.